lib/model/query/entities.js

```diff
@@ -89,9 +89,53 @@ const createSource = (details = null, subDefId = null, eventId = null) => ({ one
     .then((row) => row.id);
 };
 
+//
+// Entity event processing
+//
+
+const _createEntity = (dataset, entityData, submissionId, submissionDef, submissionDefId, event, parentEvent) => async ({ Entities, Audits }) => {
+  const partial = await Entity.fromParseEntityData(entityData);
+
+  const sourceDetails = { submission: { instanceId: submissionDef.instanceId }, parentEventId: parentEvent ? parentEvent.id : undefined };
+  const sourceId = await Entities.createSource(sourceDetails, submissionDefId, event.id);
+  const entity = await Entities.createNew(dataset, partial, submissionDef, sourceId);
+
+  return Audits.log({ id: event.actorId }, 'entity.create', { acteeId: dataset.acteeId },
+    {
+      entityId: entity.id, // Added in v2023.3 and backfilled
+      entityDefId: entity.aux.currentVersion.id, // Added in v2023.3 and backfilled
+      entity: { uuid: entity.uuid, dataset: dataset.name },
+      submissionId,
+      submissionDefId
+    });
+};
+
+const _updateEntity = (dataset, entityData, submissionId, submissionDef, submissionDefId, event, parentEvent) => async ({ Entities }) => {
+
+  // Get version of entity on the server
+  const serverEntity = await Entities.getById(dataset.id, entityData.system.id).then(o => o.get());
+
+  // merge data
+  const mergedData = mergeRight(serverEntity.aux.currentVersion.data, entityData.data);
+  const mergedLabel = entityData.system.label || serverEntity.aux.currentVersion.label;
+
+  // make some kind of source object
+  const sourceDetails = {
+    submission: { instanceId: submissionDef.instanceId },
+    dataParsedFromSubmission: entityData.data, // or some other info
+    labelParsedFromSubmission: entityData.label,
+    baseVersionParsedFromSubmission: 0,
+    parentEventId: parentEvent ? parentEvent.id : undefined,
+  };
+  const sourceId = await Entities.createSource(sourceDetails, submissionDefId, event.id);
+
+  // TODO: createVersion infers creator id from context.auth, shoving the submitter id in as another arg
+  // should be refactored
+  return Entities.createVersion(dataset, serverEntity, mergedData, mergedLabel, serverEntity.aux.currentVersion.version + 1, sourceId, null, submissionDef.submitterId);
+};
 
 // Entrypoint to where submissions (a specific version) become entities
-const _processSubmissionEvent = (event, parentEvent) => async ({ Datasets, Entities, Submissions, Forms, Audits }) => {
+const _processSubmissionEvent = (event, parentEvent) => async ({ Datasets, Entities, Submissions, Forms }) => {
```

Comment:
(Not sure exactly where to put this comment.) An entity update should be processed even if the entity list requires approval for entity _creation_. So I think this PR should update this `if`:

https://github.com/getodk/central-backend/blob/6f7750b19f95e34b6b20f5f13891f6fb027cd393/lib/model/query/entities.js#L143-L145

Maybe that `if` should be moved to `_createEntity()`.

Reply:
will do in #1013! because of the way i've split things up, it's hard for me to think about where to put these changes. maybe i shouldn't have split it into two PRs...